src/io/XTcpSocket.cpp

```diff
@@ -0,0 +1,193 @@
+#include "XTcpSocket.h"
+
+// platform dependent system headers
+#ifdef _MSC_VER
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#define SHUT_RDWR SD_BOTH
+#define MSG_NOSIGNAL 0
+#else
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+#include <unistd.h>
+#endif
+
+// platform independent system headers
+#include <errno.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+void XTcpSocket::FormatLastErrorStr(const char* func) {
+#ifdef _MSC_VER
+    int err;
+
+    char msgbuf[256];  // for a message up to 255 bytes.
+
+    msgbuf[0] = '\0';  // Microsoft doesn't guarantee this on man page.
+
+    err = WSAGetLastError();
+
+    FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,  // flags
+                  NULL,                                                        // lpsource
+                  err,                                                         // message id
+                  MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),                   // languageid
+                  msgbuf,                                                      // output buffer
+                  sizeof(msgbuf),                                              // size of msgbuf, bytes
+                  NULL);                                                       // va_list of arguments
+
+    if (!*msgbuf)
+        snprintf(lastError, 255, "%s error : %d ", err);  // provide error # if no string available
+    else
+        snprintf(lastError, 255, "%s error : %s ", func, msgbuf);
+
+#else
+    snprintf(lastError, 255, "%s error : %s ", func, strerror(errno));
+#endif
+}
+
+const char* XTcpSocket::getLastErrorStr(void) {
+    return lastError;
+}
+
+void XTcpSocket::close() {
+    // shutdown (sockfd,SHUT_RDWR );
+#ifdef _MSC_VER
+    closesocket(sockfd);
+#else
+    ::close(sockfd);
+#endif
+    sockfd = -1;
+}
+
+void XTcpSocket::abort() {
+    close();
+}
+
+int XTcpSocket::write(const QByteArray& buff, const size_t size) {
+    int ret = send(sockfd, buff.data(), size, MSG_NOSIGNAL);
+    if (ret != (int)size) {
```

Comment:
please use C++ style cases (`static_cast`). This also goes for all of the `struct` casts further down.

Reply:
Replaced.